### PR TITLE
Added tidycensus dataset

### PR DESCRIPTION
**Pull request by Aakash3101**
Python script for installing the datasets present in the R package `tidycensus` using rpy2.

Some other changes that I made after yesterday's meeting, are to check if the raw data files are present or not. If they are present use those files and install them.

### CODE REVIEW

**Comment by Aakash3101** [comment]
Okay, so the tests are failing because when the scripts are read, it reads the import lines, and it could not find the `rpy2` package. So that is the reason why the tests fail. And I have tested that only the import lines are read, the script is actually not executed so we can put the import statements in a try-except block like:
``` python
try:
    import rpy2.robjects as ro
    import rpy2.robjects.packages as rpackages
    from rpy2.robjects import pandas2ri
    from rpy2.robjects.conversion import localconverter
except ImportError:
    pass
```

File: scripts/tidycensus.py [hunk 1]
```diff
@@ -0,0 +1,220 @@
...
+        self.citation = ""
+        self.licenses = [{"name": "MIT"}]
+        self.description = "An integrated R interface to the decennial US Census and American Community Survey APIs and" \
+                           "the US Census Bureau's geographic boundary files. Allows R users to return Census and ACS data as" \
+                           "tidyverse-ready data frames, and optionally returns a listcolumn with feature geometry for all Census" \
+                           "geographies."
+        self.keywords = ["census", "US", "tidycensus", "R", "geographic boundaries"]
+        self.encoding = 'utf-8'
+
+    def census_api_key(api_key):
```
Comment (henrykironde) [inline scripts/tidycensus.py:46]: Add a doc string

File: scripts/tidycensus.py [hunk 2]
```diff
@@ -0,0 +1,220 @@
...
+        library(tidycensus)
+        census_api_key("{}", install = TRUE)
+        home <- Sys.getenv("HOME")
+        renv <- file.path(home, ".Renviron")
+        readRenviron(renv)
+        '''.format(api_key)
+
+        ro.r(r_string)
+
+    def install_tidycensus():
```
Comment (henrykironde) [inline scripts/tidycensus.py:57]: Add a doc string

File: scripts/tidycensus.py [hunk 3]
```diff
@@ -0,0 +1,220 @@
...
+                    print(output)
+                    print(
+                        "*Please install the system dependencies for {} package*".format(
+                            package))
+                    return False
+
+                else:
+                    print("Package {} is installed".format(package))
+            else:
+                print("Package {} is installed".format(package))
```
Comment (henrykironde) [inline scripts/tidycensus.py:71]: These two print statements seem like we could just remove them one.

File: scripts/tidycensus.py [hunk 4]
```diff
@@ -0,0 +1,220 @@
...
+
+        ro.r(r_string)
+
+    def install_tidycensus():
+        for package in packages:
+            if not rpackages.isinstalled(package):
+                output = subprocess.getoutput(R_string.format(package_name=package))
+                if bool(re.search(pattern, output)):
+                    print(output)
+                    print(
```
Comment (henrykironde) [inline scripts/tidycensus.py:63]: Could you comment a sample run when the package is 
1) Not available so it is installed
2) Run it with out an internet connection am curious to see what we get

File: scripts/tidycensus.py [hunk 5]
```diff
@@ -0,0 +1,220 @@
...
+                            package))
+                    return False
+
+                else:
+                    print("Package {} is installed".format(package))
+            else:
+                print("Package {} is installed".format(package))
+        return True
+
+    def download_raw_data(self, datasets):
```
Comment (henrykironde) [inline scripts/tidycensus.py:74]: Add a doc string, and may be a comment on the if statements to explain why we are using the if statement.

File: scripts/tidycensus.py [hunk 6]
```diff
@@ -0,0 +1,220 @@
...
+            'fips_codes.csv', 'mig_recodes.csv', 'pums_variables.csv',
+            'state_laea/state_laea.shp', 'county_laea/county_laea.shp'
+        ]
+        for filepath in filepaths:
+            if os.path.isfile(os.path.join(data_path, filepath)):
+                filename, file_ext = os.path.splitext(os.path.basename(filepath))
+                file_paths[filename] = os.path.join(data_path, filepath)
+                download = False
+
+        if (key[0]
```
Comment (henrykironde) [inline scripts/tidycensus.py:147]: can be `if not key[0] and ...`
This part `[x for x in packages if rpackages.isinstalled(x)` looks like it checking if its installed the package, and then we call "self.install_tidycensus()" which also checks the same. 

Could we organize this in a way that we first check and make sure all libraries are installed, and the api keys is present, then we do not have to check several times.

File: scripts/tidycensus.py [hunk 7]
```diff
@@ -0,0 +1,220 @@
...
+                == '') and (len([x for x in packages if rpackages.isinstalled(x)]) > 0):
+            result = self.install_tidycensus()
+            print("Installing tidycensus...")
+            if result:
+                print("Installation completed!!!")
+                print(
+                    '*The API key provided to you from the Census formated in quotes. A key can be acquired at http://api.census.gov/data/key_signup.html*'
```
Comment (henrykironde) [inline scripts/tidycensus.py:154]: Avoid very long lines, you have some tools that you could use to do some refactoring like yapf. But be care full of applying it to scripts. Especially the the very first lines should not be changed.
```python
# -*- coding: utf-8 -*-
#retriever
```

File: scripts/tidycensus.py [hunk 8]
```diff
@@ -0,0 +1,220 @@
...
+                exit()
+            self.download_raw_data(datasets)
+
+        elif (key[0] == '') and (len([x for x in packages if rpackages.isinstalled(x)])
+                                 == 0):
+            print(
+                '*The API key provided to you from the Census formated in quotes. A key can be acquired at http://api.census.gov/data/key_signup.html*'
+            )
+            print("Enter your census api key here : ", end="")
+            api_key = input()
```
Comment (henrykironde) [inline scripts/tidycensus.py:170]: We should not ask people to enter any information, they should use config files for keys

**Review by henrykironde — COMMENTED** [review]
Over all, seems good. I have made some comments.
Please include some example commands that you have run and screenshots  where applicable.

**Comment by henrykironde** [comment]
@Aakash3101, let us work on the clean up together, I have got some new designs we can follow.

**Comment by Aakash3101** [comment]
Sure

**Comment by Aakash3101** [comment]
Installing tidycensus without internet:

Screenshot from 2021-08-10 23-55-46

---

Installing tidycensus with internet:

Screenshot from 2021-08-10 23-58-54

**Comment by Aakash3101** [comment]
@henrykironde could you share your changes in the `install_tidycensus` function

**Comment by Aakash3101** [comment]
@henrykironde I am unable to make changes in the script for it to follow the `retriever download tidycensus --path <PATH>` command. The raw data is getting saved in the `~/.retriever/raw_data/tidycensus` directory and not at the specified `PATH`.

I think @kkothari2001 would be able to help me here, and he would also achieve his task from this.

**Pull request merged**